### PR TITLE
Conda python pdm refactor

### PR DESCRIPTION
**Pull request by griff-rees**
Replace as much as possible of `environments.yml` with `python/pyproject.toml` using [`pdm`](https://pdm-project.org/latest/).

### CODE REVIEW

File: .conda-linux-64.lock [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+https://conda.anaconda.org/conda-forge/linux-64/tiledb-2.20.1-h6997b71_2.conda#0699e6cbb8521c6cab63b6d1f8de7f74
+https://conda.anaconda.org/conda-forge/noarch/cachecontrol-0.14.0-pyhd8ed1ab_0.conda#a661c39e223bf3038b38126b0bbf43d9
+https://repo.anaconda.com/pkgs/main/linux-64/ensureconda-1.4.3-py311h06a4308_0.conda#2857f498b7ed0188cca694d05fc95a5e
+https://repo.anaconda.com/pkgs/main/linux-64/folium-0.14.0-py311h06a4308_0.conda#ccb8ca695c6450a757a7c0221f02ff10
+https://conda.anaconda.org/conda-forge/linux-64/libgdal-3.8.4-h9323651_0.conda#f0444ecc68c3f7d0855c9dd6bc3424a7
+https://repo.anaconda.com/pkgs/main/linux-64/nbconvert-7.10.0-py311h06a4308_0.conda#45765508cfa84c24b7faba9ce063ea83
+https://repo.anaconda.com/pkgs/main/linux-64/requests-toolbelt-1.0.0-py311h06a4308_0.conda#53b9b999d06502302fcabedc60184654
+https://conda.anaconda.org/conda-forge/noarch/unearth-0.14.0-pyhd8ed1ab_0.conda#6ee75538d26bba9f6b5c7e4043d6562a
+https://conda.anaconda.org/conda-forge/noarch/cachecontrol-with-filecache-0.14.0-pyhd8ed1ab_0.conda#4c08fa6e7d1d3f124ad815e21b2210e9
+https://conda.anaconda.org/conda-forge/linux-64/gdal-3.8.4-py311h8be719e_0.conda#de4d128805c8e1cbe408a93be1f715f3
```
Comment (andrewphilipsmith) [inline .conda-linux-64.lock:347]: GDAL==3.8.4 here

File: environment.yml [hunk 2]
```diff
@@ -4,43 +4,15 @@ channels:
   - defaults
   - conda-forge
 dependencies:
-  - llvm-openmp=17.0.6
-  - python=3.10
+  # - llvm-openmp=17.0.6
+  - python=3.11
   - gdal=3.8.3
```
Comment (andrewphilipsmith) [inline environment.yml:9]: GDAL==3.8.3 here

File: python/pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    "tqdm<5.0.0,>=4.66.2",
+    "scipy<2.0.0,>=1.12.0",
+    "rioxarray<1.0.0,>=0.15.1",
+    "rasterio<2.0.0,>=1.3.9",
+    "python-dateutil<3.0.0,>=2.8.2",
+    "jupyterlab<5.0.0,>=4.1.2",
+    "matplotlib>=3.8.3",
+    "geopandas>=0.14.3",
+    "openpyxl>=3.1.2",
+    "GDAL>=3.8.4",
```
Comment (andrewphilipsmith) [inline python/pyproject.toml:33]: GDAL==3.8.4 here

File: .conda-linux-64.lock (hunk 1 above)
Comment (griff-rees) [reply .conda-linux-64.lock]: Thanks. I've shifted this back to `GDAL==3.8.3`.

File: python/pyproject.toml (hunk 3 above)
Comment (griff-rees) [reply python/pyproject.toml]: As above

**Comment by andrewphilipsmith** [comment]
I ran the tests successfully in Docker. Happy for this to be merged.

File: python/clim_recal/resample.py
[the remark this answers is not in the document]
Comment (griff-rees) [reply python/clim_recal/resample.py:1]: Cool, till then

File: python/pyproject.toml
[the remark this answers is not in the document]
Comment (griff-rees) [reply python/pyproject.toml:1]: Hoping that's a *good* thing, and either way, of course happy to address anything more via a pull request to `main`.

File: python/conftest.py
[the remark this answers is not in the document]
Comment (griff-rees) [reply python/conftest.py:1]: Till that pull request...

**Pull request merged**